src/compute/common/content_size.rs

```diff
@@ -1,32 +1,84 @@
-use crate::geometry::{Point, Size};
+use crate::geometry::{Point, Rect, Size};
 use crate::style::Overflow;
-use crate::util::sys::f32_max;
-
-#[inline(always)]
-/// Determine how much width/height a given node contributes to it's parent's content size
-pub(crate) fn compute_content_size_contribution(
-    location: Point<f32>,
-    size: Size<f32>,
-    content_size: Size<f32>,
-    overflow: Point<Overflow>,
-) -> Size<f32> {
-    let size_content_size_contribution = Size {
-        width: match overflow.x {
-            Overflow::Visible => f32_max(size.width, content_size.width),
-            _ => size.width,
-        },
-        height: match overflow.y {
-            Overflow::Visible => f32_max(size.height, content_size.height),
-            _ => size.height,
-        },
-    };
-    if size_content_size_contribution.width > 0.0 && size_content_size_contribution.height > 0.0 {
-        Size {
-            width: location.x + size_content_size_contribution.width,
-            height: location.y + size_content_size_contribution.height,
-        }
-    } else {
-        Size::ZERO
+
+pub(crate) fn expand_scrollable_overflow(
+    container_scrollable_overflow: &mut Rect<f32>,
+    container_scroll_origin: Point<f32>,
+
+    item_location: Point<f32>,
+    item_size: Size<f32>,
+    item_border: Rect<f32>,
+    item_scrollbar_gutter: Size<f32>,
+    item_overflow: Point<Overflow>,
+    item_scrollable_overflow: Rect<f32>,
+) {
+    let item_rect = Rect::from_top_left_and_size(item_location, item_size);
+
+    // From CSS Overflow 3, section 2.2:
+    //   The scrollable overflow area is the union of:
+    //     ...
+    //   - The border boxes of all boxes for which it is the
+    //     containing block and whose border boxes are positioned not
+    //     wholly in the unreachable scrollable overflow region,
+    //     accounting for transforms by projecting each box onto the
+    //     plane of the element that establishes its 3D rendering
+    //     context. [CSS3-TRANSFORMS]
+    if item_rect.right < container_scroll_origin.x || item_rect.bottom < container_scroll_origin.y {
+        return;
+    }
+
+    // From CSS Overflow 3, section 2.2:
+    //   The scrollable overflow area is the union of:
+    //     ...
+    //     Border boxes with zero area do not affect the scrollable
+    //     overflow area.
+    if item_size.has_non_zero_area() {
+        container_scrollable_overflow.union_with(item_rect);
+    }
+
+    // From CSS Overflow 3, section 2.2:
+    //   The scrollable overflow area is the union of:
+    //     ...
+    //   - The margin areas of grid item and flex item boxes for
+    //     which the box establishes a containing block.
+
+    // TODO: open questions:
+    // - what does this mean?
+    // - is this predicated the same as the previous item?
+    // - assuming the margins are positive, don't this subsume the
+    //   previous item?
```

Comment:
Are you missing margin-box vs border-box here? This seems to be saying that margins should extend the overflow scroll contribution for flex and grid items. Whereas they would be ignored for other kinds of box.